server/src/main/java/org/spine3/server/entity/VisibilityGuard.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2017, TeamDev Ltd. All rights reserved.
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.spine3.server.entity;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Sets;
+import com.google.protobuf.Message;
+import org.spine3.annotation.Internal;
+import org.spine3.option.EntityOption.Visibility;
+import org.spine3.option.EntityOptions;
+import org.spine3.type.TypeName;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.Iterables.transform;
+import static com.google.common.collect.Maps.filterValues;
+import static com.google.common.collect.Maps.newHashMap;
+import static org.spine3.util.Exceptions.newIllegalArgumentException;
+import static org.spine3.util.Exceptions.newIllegalStateException;
+
+/**
+ * A registry of repositories that controls access to them depending on the visibility of
+ * corresponding entity states.
+ *
+ * @author Alexander Yevsyukov
+ * @see EntityOptions
+ */
+@Internal
+public final class VisibilityGuard {
+
+    private final Map<Class, RepositoryAccess> repositories = newHashMap();
+
+    private VisibilityGuard() {
+        // Prevent instantiation from outside.
+    }
+
+    /**
+     * Creates a new instance of the guard.
+     */
+    public static VisibilityGuard newInstance() {
+        return new VisibilityGuard();
+    }
+
+    /**
+     * Registers the passed repository with the guard.
+     */
+    public void register(Repository repository) {
+        checkNotNull(repository);
+        final Class stateClass = repository.getEntityStateClass();
+        checkNotAlreadyRegistered(stateClass);
+        repositories.put(stateClass, new RepositoryAccess(repository));
+    }
+
+    private void checkNotAlreadyRegistered(Class stateClass) {
```

Comment:
Let's check this parameter for `null` in order to be consistent. 

See more below on the same matter.